### PR TITLE
feat: Remove caching in wasmer_wamr mode

### PR DESCRIPTION
**Pull request by mattyg**
In `wasmer_wamr` mode, expose a public function `build_module` to build a wasm module directly, bypassing the cache.

This is a prerequisite for https://github.com/holochain/holochain/pull/4441

This PR is merging into `chore/cleanup-make-private` for a more legible diff.

### CODE REVIEW

File: crates/common/src/result.rs [hunk 1]
```diff
@@ -30,8 +30,9 @@ pub enum WasmErrorInner {
     /// AND wasm execution MUST immediately halt.
     /// The `Vec<u8>` holds the encoded data as though the guest had returned.
     HostShortCircuit(Vec<u8>),
-    /// Wasmer failed to compile machine code from wasm byte code.
-    Compile(String),
+    /// Wasmer failed to build a Module from wasm byte code.
+    /// With the feature `wasmer_sys` enabled, building a Module includes compiling the wasm.
+    ModuleBuild(String),
```
Comment (mattyg) [inline crates/common/src/result.rs:35]: Rename error type to clarify that it is an error arising when creating a new wasmer module. Only in `wasmer_sys` mode does compilation occur here.

File: crates/guest/Cargo.toml [hunk 2]
```diff
@@ -14,7 +14,7 @@ crate-type = ["cdylib", "rlib"]
 path = "src/guest.rs"
 
 [dependencies]
-holochain_serialized_bytes = { version = "=0.0.55", features = [] }
+holochain_serialized_bytes = "=0.0.55"
```
Comment (mattyg) [inline crates/guest/Cargo.toml:17]: Unrelated clean up. Cargo features are additive and this dep does not specify `no-default-features`, so I believe the `features = []` has no effect.

File: crates/host/src/module.rs [hunk 3]
```diff
@@ -52,11 +49,20 @@ pub struct InstanceWithStore {
...
+type PlruKeyMap = BiMap<usize, CacheKey>;
+/// Modules serialize to a vec of bytes as per wasmer.
+type SerializedModule = Bytes;
+
 /// Higher level trait over the plru cache to make it a bit easier to interact
 /// with consistently. Default implementations for key functions are provided.
 /// Notably handles keeping the mapping between cache keys and items, and the
 /// plru tracking including touching and evicting.
-pub trait PlruCache {
+trait PlruCache {
```
Comment (mattyg) [inline crates/host/src/module.rs:65]: Unrelated cleanup. This is only used internally.

File: crates/host/src/module.rs [hunk 4]
```diff
@@ -104,12 +110,14 @@ pub trait PlruCache {
 
     /// Delete the plru for a given cache key. Care must be taken to ensure this
     /// is not called before a subsequent call to `plru_key` or it will panic.
+    #[allow(dead_code)]
```
Comment (mattyg) [inline crates/host/src/module.rs:113]: Unrelated cleanup. This is never used, but may be in the future.

File: crates/host/src/module.rs [hunk 5]
```diff
@@ -104,12 +110,14 @@ pub trait PlruCache {
...
     fn trash(&mut self, key: &CacheKey) {
         let plru_key = self.plru_key(key);
         self.plru_mut().trash(plru_key);
     }
 
     /// Remove an item from the cache and the associated plru entry.
+    #[allow(dead_code)]
```
Comment (mattyg) [inline crates/host/src/module.rs:120]: Unrelated cleanup. This is never used, but may be in the future.

File: crates/host/src/module.rs [hunk 6]
```diff
@@ -312,7 +321,7 @@ impl SerializedModuleCache {
 /// the cache to create callable instances is slow. Therefore modules are
 /// cached in memory after deserialization.
 #[derive(Default, Debug)]
-pub struct DeserializedModuleCache {
+struct DeserializedModuleCache {
```
Comment (mattyg) [inline crates/host/src/module.rs:324]: Unrelated cleanup. This is only used internally.

File: crates/host/src/module/wasmer_sys.rs [hunk 7]
```diff
@@ -23,7 +27,7 @@ pub const WASM_METERING_LIMIT: u64 = 10_000_000;
 
 /// Generate an engine with a wasm compiler
 /// and Metering (use limits) in place.
-pub fn make_engine() -> Engine {
+pub(crate) fn make_engine() -> Engine {
```
Comment (mattyg) [inline crates/host/src/module/wasmer_sys.rs:30]: This is only used internally in this crate

File: crates/host/src/module/wasmer_sys.rs [hunk 8]
```diff
@@ -50,7 +54,7 @@ pub fn make_engine() -> Engine {
     engine
 }
 
-pub fn make_runtime_engine() -> Engine {
+pub(crate) fn make_runtime_engine() -> Engine {
```
Comment (mattyg) [inline crates/host/src/module/wasmer_sys.rs:57]: This is only used internally in this crate

File: crates/host/src/module/wasmer_wamr.rs [hunk 9]
```diff
@@ -24,3 +15,42 @@ pub fn build_ios_module(_wasm: &[u8]) -> Result<Module, CompileError> {
...
+    let module = res.map_err(|e| wasm_error!(WasmErrorInner::ModuleBuild(e.to_string())))?;
+    Ok(Arc::new(module))
+}
+
+#[cfg(test)]
+pub mod tests {
+    use super::build_module;
+
+    #[test]
+    fn build_module_test() {
```
Comment (mattyg) [inline crates/host/src/module/wasmer_wamr.rs:41]: smoke test for building a wasm module in `wasmer_wamr` mode.

File: crates/host/src/module.rs [hunk 10]
```diff
@@ -36,9 +42,9 @@ pub use wasmer_wamr::*;
 pub type CacheKey = [u8; 32];
 /// Plru uses a usize to track "recently used" so we need a map between 32 byte cache
 /// keys and the bits used to evict things from the cache.
-pub type PlruKeyMap = BiMap<usize, CacheKey>;
+type PlruKeyMap = BiMap<usize, CacheKey>;
 /// Modules serialize to a vec of bytes as per wasmer.
-pub type SerializedModule = Bytes;
+type SerializedModule = Bytes;
```
Comment (mattyg) [inline crates/host/src/module.rs:47]: Unrelated cleanup. These types are only used internally.

File: scripts/bench.sh
Comment (mattyg) [inline scripts/bench.sh:1]: These files were not executable.

File: test/benches/bench.rs
Comment (mattyg) [inline test/benches/bench.rs:1]: Clippy fixes

File: test/src/test.rs
Comment (mattyg) [inline test/src/test.rs:1]: Clippy fixes

File: crates/host/src/module/wasmer_sys.rs [hunk 11]
```diff
@@ -54,6 +54,11 @@ pub(crate) fn make_runtime_engine() -> Engine {
     Engine::headless()
 }
 
+/// Build an interpreter module from wasm bytes.
+pub fn build_module(_wasm: &[u8]) -> Result<Arc<Module>, wasmer::RuntimeError> {
+    unimplemented!("The feature flag 'wasmer_wamr' must be enabled to support building a Module directly. Please use the ModuleCache instead.");
```
Comment (mattyg) [inline crates/host/src/module/wasmer_sys.rs:59]: This is keep a consistent api so we don't need to use conditional feature flags in the consumer. 

I recognize this is silly, since the function below that claims to be for ios only, is actually a general way to build a module. I'd like to wait until we remove the prebuild ios modules fully before removing it.

**Review by ThetaSinner — APPROVED** [review]
Not sure how much value I'm providing with a review here but the reason for the change makes sense and all the code changes seem consistent with the goal and each other :)

File: crates/common/src/result.rs [hunk 12]
```diff
@@ -51,7 +52,7 @@ impl WasmErrorInner {
             // Bad memory is bad memory.
             | Self::Memory
             // Failing to compile means we cannot reuse.
```
[the remark this answers is not in the document]
Comment (mattyg) [reply crates/common/src/result.rs]: updated to `// Failing to build the wasmer Module means we cannot use it.`

File: crates/host/src/module.rs [hunk 13]
```diff
@@ -215,11 +225,12 @@ impl SerializedModuleCache {
                 // of middleware like metering. Middleware is compiled into the
                 // module once and available in all instances created from it.
                 let compiler_engine = (self.make_engine)();
-                let module = Module::from_binary(&compiler_engine, wasm)
-                    .map_err(|e| wasm_error!(WasmErrorInner::Compile(e.to_string())))?;
+                let res = Module::from_binary(&compiler_engine, wasm);
+                let module =
```
[the remark this answers is not in the document]
Comment (mattyg) [reply crates/host/src/module.rs]: Accidentally included in pr -- reverted
